cli/src/semgrep/core_output.py

```diff
@@ -191,15 +190,17 @@ def convert_to_rule_match(match: out.CoreMatch) -> RuleMatch:
             fix=fix,
         )
 
-    # TODO: Dict[out.RuleId, RuleMatchSet]
-    findings: Dict[Rule, RuleMatchSet] = {rule: RuleMatchSet(rule) for rule in rules}
-    seen_cli_unique_keys: Set[Tuple] = set()
+    by_unique_key: Dict[Tuple, RuleMatch] = {}
     for match in res.results:
-        rule = rule_table[match.check_id.value]
         rule_match = convert_to_rule_match(match)
-        if rule_match.cli_unique_key in seen_cli_unique_keys:
-            continue
-        seen_cli_unique_keys.add(rule_match.cli_unique_key)
+        curr = by_unique_key.setdefault(rule_match.cli_unique_key, rule_match)
+        if rule_match.should_report_instead(curr):
+            by_unique_key[rule_match.cli_unique_key] = rule_match
+
+    # TODO: Dict[out.RuleId, RuleMatchSet]
+    findings: Dict[Rule, RuleMatchSet] = {rule: RuleMatchSet(rule) for rule in rules}
+    for rule_match in by_unique_key.values():
+        rule = rule_table[rule_match.rule_id]
         findings[rule].add(rule_match)
```

Comment:
I love that you got rid of the complicated control flow. I am confused why the wasn't just `if not` statement to add the key.